src/common.c

```diff
@@ -371,6 +371,234 @@ np_ntf_send(struct nc_session *ncs, const struct timespec *timestamp, struct lyd
     return rc;
 }
 
+/**
+ * @brief Create common session parameters to identify a management session.
+ *
+ * Detail description is in the common-session-parms grouping located in the ietf-netconf-notifications module.
+ *
+ * @param[in] new_session Created NC session.
+ * @param[in,out] notif Notification to which the session parameters are to be added.
+ * @return 0 on success.
+ */
+static LY_ERR
+np_prepare_notif_common_session_parms(const struct nc_session *session, struct lyd_node *notif)
+{
+    char *value;
+    char num32[11]; /* max bytes for 32-bit unsigned number + \0 */
+
+    assert(session && notif);
+
+    /* create 'username' node */
+    value = (char *)nc_session_get_username(session);
+    if (lyd_new_term(notif, notif->schema->module, "username", value, 0, NULL)) {
+        return -1;
+    }
+
+    /* create 'session-id' node */
+    sprintf(num32, "%" PRIu32, nc_session_get_id(session));
+    if (lyd_new_term(notif, notif->schema->module, "session-id", num32, 0, NULL)) {
+        return -1;
+    }
+
+    /* create 'source-host' node */
+    if (nc_session_get_ti(session) != NC_TI_UNIX) {
+        value = (char *)nc_session_get_host(session);
+        if (lyd_new_term(notif, notif->schema->module, "source-host", value, 0, NULL)) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+int
+np_send_notif_session_start(const struct nc_session *new_session, sr_session_ctx_t *sr_session, uint32_t sr_timeout)
+{
+    int rc = 0, r;
+    const struct ly_ctx *ly_ctx;
+    const struct lys_module *mod;
+    struct lyd_node *notif = NULL;
+
+    ly_ctx = nc_session_get_ctx(new_session);
+    if ((mod = ly_ctx_get_module_implemented(ly_ctx, "ietf-netconf-notifications"))) {
+        /* create 'netconf-session-start' notification */
+        if (lyd_new_path(NULL, ly_ctx, "/ietf-netconf-notifications:netconf-session-start", NULL, 0, &notif)) {
+            rc = -1;
+            goto cleanup;
+        }
+
+        /* create 'common-session-parms' grouping */
+        if ((rc = np_prepare_notif_common_session_parms(new_session, notif))) {
+            goto cleanup;
+        }
+
+        /* send notification */
+        if ((r = sr_notif_send_tree(sr_session, notif, sr_timeout, 0))) {
+            WRN("Failed to send a notification (%s).", sr_strerror(r));
+            rc = -1;
+            goto cleanup;
+        }
+
+        VRB("Generated new event (netconf-session-start).");
+    }
+
+cleanup:
+    lyd_free_tree(notif);
+    return rc;
+}
+
+int
+np_send_notif_session_end(const struct nc_session *session, sr_session_ctx_t *sr_session, uint32_t sr_timeout)
+{
+    int rc = 0, r;
+    const struct ly_ctx *ly_ctx;
+    const struct lys_module *mod;
+    struct lyd_node *notif = NULL;
+    char num32[11]; /* max bytes for 32-bit unsigned number + \0 */
+    char *value;
+
+    ly_ctx = nc_session_get_ctx(session);
+    if ((mod = ly_ctx_get_module_implemented(ly_ctx, "ietf-netconf-notifications"))) {
```

Comment:
This `if` could be negated and lead to `cleanup`, would avoid having the whole function in the condition (which I prefer to not do). Also, if you have the `mod`, you should use it and call `lyd_new_inner()` instead of `lyd_new_path()`, it is unused currently.